bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/SingleColumnValueFilterAdapter.java

```diff
@@ -51,90 +57,171 @@ public SingleColumnValueFilterAdapter(ValueFilterAdapter delegateAdapter) {
+   *             qualifier: [filter.qualifier],
+   *           }
+   *         },
+   *         false_filter: { pass_all: true }
+   *       }
+   *     }
+   *   ]
+   * </pre>
+   *
+   * NOTE: This logic can also be expressed as nested predicates, but that approach creates really poor
+   * performance on the server side.
+   * <p>
+   */
   @Override
   public RowFilter adapt(FilterAdapterContext context, SingleColumnValueFilter filter)
       throws IOException {
+
+    // filter to check if the column exists
+    RowFilter columnSpecFilter = getColumnSpecFilter(
+      filter.getFamily(),
+      filter.getQualifier(),
+      filter.getLatestVersionOnly());
+
+    // filter to return the row if the condition is met
+    RowFilter emitRowsWithValueFilter = RowFilter.newBuilder()
+      .setCondition(
+          Condition.newBuilder()
+              .setPredicateFilter(
+                  RowFilter.newBuilder()
+                      .setChain(
+                          columnSpecFilter.getChain().toBuilder()
+                              .addFilters(createValueMatchFilter(context, filter))
+                              .build()))
+              .setTrueFilter(ALL_VALUES_FILTER))
+      .build();
+
     if (filter.getFilterIfMissing()) {
-      return createEmitRowsWithValueFilter(context, filter);
+      return emitRowsWithValueFilter;
     } else {
-      return RowFilter.newBuilder()
-          .setCondition(
-              Condition.newBuilder()
-                  .setPredicateFilter(createColumnSpecFilter(filter))
-                  .setTrueFilter(createEmitRowsWithValueFilter(context, filter))
-                  .setFalseFilter(ALL_VALUES_FILTER))
-          .build();
+      return RowFilter.newBuilder().setInterleave(
+        Interleave.newBuilder()
+          .addFilters(emitRowsWithValueFilter)
+          .addFilters(RowFilter.newBuilder()
+            .setCondition(
+                Condition.newBuilder()
+                    .setPredicateFilter(columnSpecFilter)
+                    .setFalseFilter(ALL_VALUES_FILTER))
+            .build())
+          .build()
+      ).build();
     }
   }
 
-  /**
-   * Create a filter that will match a given family, qualifier, and cells per qualifier.
-   */
-  private RowFilter createColumnSpecFilter(SingleColumnValueFilter filter) throws IOException {
-    return RowFilter.newBuilder()
-        .setChain(Chain.newBuilder()
-            .addFilters(RowFilter.newBuilder()
-                .setFamilyNameRegexFilter(
-                    Bytes.toString(quoteRegularExpression(filter.getFamily()))))
-            .addFilters(RowFilter.newBuilder()
-                .setColumnQualifierRegexFilter(
-                    ByteStringer.wrap(quoteRegularExpression(filter.getQualifier()))))
-            .addFilters(createVersionLimitFilter(filter)))
-        .build();
-  }
+  @VisibleForTesting
+  static RowFilter getColumnSpecFilter(byte[] family, byte[] qualifier, boolean latestVersionOnly)
```

Comment:
nit: Maybe return the Chain.Builder to avoid the toBuilder().addFilters().build() step?

Reply:
The Chain needs to be reused in the `!filter.getFilterIfMissing()` case with only the family/qualifier/timestamp check and no value check.  We have to build the chain twice.